File: kor-common/src/main/java/com/sefford/kor/repositories/MemoryJsonDataSource.java
package com.sefford.kor.repositories;

import com.google.gson.Gson;
import com.sefford.kor.repositories.interfaces.RepoElement;
import com.sefford.kor.repositories.interfaces.Repository;

import java.util.*;

/**
 * Created by sefford on 6/5/17.
 */

public class MemoryJsonDataSource<K, V extends RepoElement<K>> implements Repository<K, V> {

    final Gson gson;
    final Class<V> clazz;
    final Map<K, String> cache;

    public MemoryJsonDataSource(Gson gson, Class<V> clazz) {
        this.gson = gson;
        this.clazz = clazz;
        this.cache = new HashMap<>();
    }

    @Override
    public void clear() {
        cache.clear();
    }

    @Override
    public boolean contains(K id) {
        return cache.containsKey(id);
    }

    @Override
    public void delete(K id, V element) {
        cache.remove(id);
    }

    @Override
    public void deleteAll(List<V> elements) {
        for (V element : elements) {
            delete(element.getId(), element);
        }
    }

    @Override
    public V get(K id) {
        return gson.fromJson(cache.get(id), clazz);
    }

    @Override
    public Collection<V> getAll(Collection<K> ids) {
        final List<V> elements = new ArrayList<>();
        for (K id : ids) {
            if (cache.containsKey(id)) {
                elements.add(get(id));
            }
        }
        return elements;
    }

    @Override
    public Collection<V> getAll() {
        final List<V> elements = new ArrayList<>();
        for (K id : cache.keySet()) {
            elements.add(get(id));
        }
        return elements;
    }

    @Override
    public V save(V element) {
        cache.put(element.getId(), gson.toJson(element));
        return element;
    }

    @Override
    public Collection<V> saveAll(Collection<V> elements) {
        for (V element : elements) {
            save(element);
        }
        return elements;
    }

    @Override
    public boolean isAvailable() {
        return cache != null;
    }
}

File: kor-common/src/main/java/com/sefford/kor/repositories/MemoryJsonDataSource.kt
package com.sefford.kor.repositories

import arrow.core.*
import com.chicisimo.data.StubDataSource
import com.sefford.kor.interactors.RepositoryError
import com.sefford.kor.repositories.interfaces.JsonConverter
import com.sefford.kor.repositories.interfaces.RepoElement
import com.sefford.kor.repositories.interfaces.Repository
import java.util.*

/**
 * Created by sefford on 6/5/17.
 */

class MemoryJsonDataSource<K, V : RepoElement<K>>(internal val converter: JsonConverter<V>) : Repository<K, V>, StubDataSource<K, V> {

    internal val cache: MutableMap<K, String> = mutableMapOf()

    override val all: Collection<V>
        get() {
            val elements = ArrayList<V>()
            cache.keys.forEach { get(it).map { elements.add(it) } }
            return elements
        }

    override val isReady: Boolean
        get() = true

    override fun clear() {
        cache.clear()
    }

    override fun contains(id: K): Boolean {
        return cache.containsKey(id)
    }

    override fun delete(id: K, element: V) {
        delete(id)
    }

    override fun delete(id: K) {
        cache.remove(id)
    }

    override fun get(id: K): Either<RepositoryError, V> {
        if (!cache.containsKey(id)) {
            return Either.left(RepositoryError.NotFound(id))
        }
        return converter.deserialize(cache[id])
    }

    override fun save(element: V): Either<RepositoryError, V> {
        return converter.serialize(element).fold({ Left(it) }, {
            cache[element.id] = it
            Right(element)
        })
    }
}
